File: java-snake-2d/src/main/java/metacake/snake/state/ScoreState.java
package metacake.snake.state;

import io.metacake.core.output.RenderingInstructionBundle;
import io.metacake.core.process.ActionRecognizerPipe;
import io.metacake.core.process.state.EndState;
import io.metacake.core.process.state.VoidState;
import io.metacake.s2d.output.drawing.DrawingDevice;
import io.metacake.s2d.output.drawing.instructions.DrawInstruction;
import io.metacake.s2d.output.drawing.instructions.PlaceInstruction;
import io.metacake.s2d.output.drawing.instructions.RectangleInstruction;
import io.metacake.s2d.output.drawing.instructions.TextInstruction;
import metacake.snake.entity.Snake;

import java.awt.*;

import static metacake.snake.SnakeApp.HEIGHT;
import static metacake.snake.SnakeApp.WIDTH;

public class ScoreState extends VoidState {

    private static final DrawInstruction EMPTY_SCENE = new RectangleInstruction(WIDTH, HEIGHT, Color.WHITE);
    private static final Font font = new Font("Arial", Font.PLAIN, 16);

    private int score;
    public ScoreState(Snake snake) {
        this.score = snake.size();
    }

    @Override
    public void update(long delta, ActionRecognizerPipe recognizers) {
        setTransition(EndState.endWith(this));
    }

    @Override
    public RenderingInstructionBundle renderingInstructions() {
        RenderingInstructionBundle bundle = new RenderingInstructionBundle();
        DrawInstruction withScore = new PlaceInstruction(new TextInstruction("Game Over!", font, Color.BLACK), 250, 250, EMPTY_SCENE);
        DrawInstruction image = new PlaceInstruction(new TextInstruction("Your score is " + score, font, Color.BLACK), 250, 300, withScore);
        return bundle.add(DrawingDevice.NAME(), image);
    }
}

File: snake-scala-2d/src/main/scala/metacake/scala/snake/state/ScoreState.scala
package metacake.scala.snake.state

import java.awt.{Color, Font}

import io.metacake.core.output.RenderingInstructionBundle
import io.metacake.core.process.state.GameState.Kind
import io.metacake.core.process.state.{EndState, GameState}
import io.metacake.core.process.{ActionRecognizerPipe, Transition}
import io.metacake.s2d.output.drawing.DrawingDevice
import io.metacake.s2d.output.drawing.instructions.{DrawInstruction, PlaceInstruction, RectangleInstruction, TextInstruction}
import metacake.scala.snake.SnakeApp
import metacake.scala.snake.entity.Snake
import metacake.scala.snake.state.ScoreState._

object ScoreState {
  val FONT: Font = new Font("Arial", Font.PLAIN, 16)
  val GAME_OVER: DrawInstruction = new TextInstruction("Game Over!", FONT, Color.BLACK)
  val SCENE: DrawInstruction = new PlaceInstruction(GAME_OVER, 250, 250, new RectangleInstruction(SnakeApp.WIDTH, SnakeApp.HEIGHT, Color.WHITE))
}

class ScoreState(snake: Snake) extends RenderableState {
  val score: Int = snake.segments.length

  override def tick(delta: Long, pipe: ActionRecognizerPipe): Transition = Transition.to(EndState.end())

  def render(): RenderingInstructionBundle = {
    val bundle: RenderingInstructionBundle = new RenderingInstructionBundle()
    bundle.add(DrawingDevice.NAME, new PlaceInstruction(new TextInstruction("Your score is " + score, FONT, Color.BLACK), 250, 300, SCENE))
  }

  override def kind(): Kind = Kind.NORMAL
}